### PR TITLE
Factors as Val types

### PR DESCRIPTION
**Pull request by chriselrod**
Should yield better run-time performance.
Because the kernels are probably the most costly part to compile, reducing the number needed should yield better compile time performance as well.

Figured @MasonProtter may want to look at this before merging. 

I also fixed a problem in the tests. They pass.

### CODE REVIEW

**Comment by MasonProtter** [comment]
Right, good idea!

**Comment by MasonProtter** [comment]
Any other changes you want to make or should I merge?

**Comment by chriselrod** [comment]
Single threaded Gaius allocates again, so I'm looking into that.

**Comment by MasonProtter** [comment]
I'm not seeing allocations on a single thread with this change. 

Are you sure you're running single threaded? With the new master we start threading much earlier.

**Comment by chriselrod** [comment]
Ah, you're right. I need to recompile Gaius to switch between threaded and not threaded.
I think we can come up with a better API for that.

**Comment by chriselrod** [comment]
I set the larger block size for AVX512 to 64.

Note that LoopVectorization uses an inner 32 x 4 kernel for AVX512 and an inner 12 x 4 kernel for AVX2:
```julia
julia> AmulBq2 = :(for m ∈ 1:M, n ∈ 1:N
                   C[m,n] = zero(eltype(B))
                   for k ∈ 1:K
                   C[m,n] += A[m,k] * B[k,n]
                   end
                   end);

julia> lsAmulB2 = LoopVectorization.LoopSet(AmulBq2);

julia> LoopVectorization.choose_order(lsAmulB2)
([:n, :m, :k], :m, 4, 4)
```
This means it chooses to use a 4 x 4 kernel across `n` and `m` (first 2 symbols in the vector). The vector width is 8, and `m` (second element in the tuple) is vectorized, giving us 32 x 4.
With AVX (no 512), it returns 3 x 4, because aside from the floating point registers being smaller, there also aren't as many.
We may thus expect that multiples of these sizes are optimal.

It would also be nice to add support for transposed matrices (in a separate PR). This wouldn't require any changes to the kernels, since LoopVectorization takes care of that. However, it would require a change to the PointerMatrices. The simplest thing would probably be to interface with the stridedpointers more directly. I think this can actually simplify the code.

**Comment by MasonProtter** [comment]
It looks like the latest tagged version of LoopVectorization is causing the test problem, but the master works.

**Comment by chriselrod** [comment]
I'm using some inline asm now because LLVM does something slightly suboptimal. It'll probably be fixed within the next few LLVM versions, but handling it myself also seems to speed up running LoopVectorization's test suite by a little (seems to have a bigger impact on LLVM's compile time than the resulting runtime).

However, it was accidentally getting run on an architecture that doesn't support it [causing a crash](https://travis-ci.com/chriselrod/LoopVectorization.jl/jobs/292773778). I'll release everything shortly now that I added an explicit check for `FMA3` to guard the specific asm definitions.

**Comment by MasonProtter** [comment]
Looks good to me. Ready to merge?

**Comment by chriselrod** [comment]
Okay. Note that LoopVectorization will cause problems on those architectures for a little while. I'll fix it within a couple hours, but it takes time to release a chain of packages (VectorizationBase -> SIMDPirates -> LoopVectorization).

**Comment by MasonProtter** [comment]
> Okay. Note that LoopVectorization will cause problems on those architectures for a little while. I'll fix it within a couple hours, but it takes time to release a chain of packages (VectorizationBase -> SIMDPirates -> LoopVectorization).

Those problems are independent of this PR being merged though, right? I was seeing the same problem on the master branch.

**Comment by MasonProtter** [comment]
@chriselrod I just noticed that you changed it to 
```const DEFAULT_BLOCK_SIZE = REGISTER_SIZE == 64 ? 64 : 32```
This actually gives a pretty big performance hit on non-avx512 machines. I found `64` was a good choice on non-avx512 hardware. Perhaps you wanted to do 
```julia
const DEFAULT_BLOCK_SIZE = REGISTER_SIZE == 64 ? 96 : 64
```
?

File: src/matmul.jl [hunk 1]
```diff
@@ -20,8 +20,8 @@ function mul!(C::MatTypes{T}, A::MatTypes{T}, B::MatTypes{T};
               block_size = nothing, sizecheck=true) where {T <: Eltypes}
     sizecheck && check_compatible_sizes(C, A, B)
     if isnothing(block_size)
-        if size(C, 1) >= 72
-            block_size = 48
+        if 2size(C, 1) >= 3DEFAULT_BLOCK_SIZE
+            block_size = DEFAULT_BLOCK_SIZE
```
Comment (chriselrod) [inline src/matmul.jl]: @MasonProtter The `64 : 32` was a mistake. I meant it to be `64 : 48`. Does 64 work better on avx as well?

Comment (MasonProtter) [reply src/matmul.jl]: `64` is best on AVX 512.

**Pull request merged**